Phase04

Melikad opened the pull request.

Melikad committed:
Add files via upload

initial code

Melikad committed:
bugs fixed - proper output

rahaahmadi committed:
Remove extra imports

rahaahmadi committed:
Updated

rahaahmadi committed:
Refactor Student.cs

rahaahmadi committed:
Refactor

rahaahmadi committed:
Updated

rahaahmadi committed:
Added files

rahaahmadi committed:
Used generic

nedataghizadeh79 commented on phase04/Student.cs:
```diff
@@ -0,0 +1,37 @@
+        }
+
+        public void CalculateGPA()
+        {
```
میتوانستید برای زیبایی بیشتر از توابع آماده برای یافتن معدل استفاده نمایید.

hoseinaghaei commented on .vscode/launch.json:
```diff
@@ -0,0 +1,26 @@
+{
+    "version": "0.2.0",
```
این باید بره تو گیت ایگنور

hoseinaghaei commented on .vscode/tasks.json:
```diff
@@ -0,0 +1,42 @@
+{
+    "version": "2.0.0",
```
اینم باید بره تو گیت ایگنور

hoseinaghaei commented on phase04/DataManager.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace phase04
+{
+    class DataManager
```
اسم بهتر

hoseinaghaei commented on phase04/DataManager.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    class DataManager
+    {
+        public List<T> DeserializeObject<T>(string json)
```
اسم ورودی بهتر

hoseinaghaei commented on phase04/FileReader.cs:
```diff
@@ -0,0 +1,11 @@
+    {
+        public string ReadData(string path) {
+            return File.ReadAllText(path);
+        }
```
بار ۱۰۰۰ام
کار با 
ّIO
میکنین همیشه اکسپشن هارو پیش بینی کنین

hoseinaghaei commented on phase04/Grade.cs:
```diff
@@ -0,0 +1,15 @@
+        public int StudentNumber {get; set;}
+        public string Lesson {get; set;}
+        public double Score {get; set;}
+        public Grade(int studentNumber, string lesson, double score)
```
چون کلاس قراره توسط 
json deserializer
ساخته بشه نیاز نیست کانستراکتور بذارین

hoseinaghaei commented on phase04/Program.cs:
```diff
@@ -0,0 +1,15 @@
+            FileReader fileReader = new FileReader();
+            DataManager dataManager = new DataManager();
+            
+            organizer.Run(fileReader, dataManager);
```
بهتره تابع ران خروجی داشته باشه

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+        public void GetData(FileReader fileReader, DataManager dataManager)
+        {
+            string studentsJsonInput = fileReader.ReadData("Students.json");
```
ادرس رو هارد کد ندین

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            return StudentsList.Where(Student => Student.StudentNumber == studentNumber).First();
+        }
+        public void GetData(FileReader fileReader, DataManager dataManager)
```
اینجا برید برای دوتا کلاس ورودی اینترفیس بذارین
چون برای تست باید ماک بشن

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+        
+        public Student FindStudent(int studentNumber)
+        {
+            return StudentsList.Where(Student => Student.StudentNumber == studentNumber).First();
```
تو خود تابع 
First
میتونید شرط بذارین

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+            string scoresJsonInput = fileReader.ReadData("Scores.json");
+            ScoresList = dataManager.DeserializeObject<Grade>(scoresJsonInput);   
+        }
+        public void PrintOutput()
```
کار 
IO
رو به یک اینترفیس جداگانه بدین
به کلاس 
ScoresOrganizer
ربطی نداره پرینت کردن

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+        public void SortStudentsList()
+        {
+            StudentsList.Sort((a, b) => b.GPA.CompareTo(a.GPA));
```
از تابع
OrderBy
لینک استفاده کنید

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+        public void PrintOutput()
+        {
+            for(int i = 0; i < 3; i++)
```
با کمک لینک بعد از سورت کردن با استفاده از تابع 
Take
میتونید هرتعدادی که خواستین بگیرین
و نیاز نیست دیه فور با i بزنین

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            StudentsList.Sort((a, b) => b.GPA.CompareTo(a.GPA));
+        }
+        public void AddGrades()
```
اسم بد

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            foreach(Grade grade in ScoresList)
+            {
+                Student student = FindStudent(grade.StudentNumber);
```
برید راجج
Join & GroupJoin
تو لینک بخونین
و سعی کنین با کمک اون بزنید کدتون رو

hoseinaghaei commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,56 @@
+                student.AddScore(grade.Score);
+            }
+        }
+        public void SetGPAs()
```
اسم درست تر بذارین

hoseinaghaei commented on phase04/Student.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace phase04 
+{
+    class Student
```
کلا ساختار کلاس بده
وقتی یک شی رو دیسریالاز میکنین باید همون فیلدهای تو جیسون رو فقط براش بذارین
الان خوانایی کد خیلی بده

hoseinaghaei commented on phase04/bin/Debug/net5.0/phase04.deps.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "runtimeTarget": {
```
گیت ایگنور

hoseinaghaei commented on phase04/bin/Debug/net5.0/phase04.runtimeconfig.dev.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "runtimeOptions": {
```
گیت ایگنور
کلی فایل 
dll
دیگم پوش کردین
برین درستش کنین

hoseinaghaei commented on phase04/obj/Debug/net5.0/phase04.AssemblyInfo.cs:
```diff
@@ -0,0 +1,23 @@
+//------------------------------------------------------------------------------
+// <auto-generated>
```
بجز کلاس های 
.cs و 
.csproj
و دوتا جیسون دیتا همه چی باید بره تو گیت ایگنور

rahaahmadi committed:
Delete .vscode directory

rahaahmadi committed:
Changed student class

rahaahmadi committed:
Changed Scores class

rahaahmadi committed:
Changed Files

rahaahmadi committed:
Delete DataManager.cs

rahaahmadi committed:
Delete phase04/obj directory

rahaahmadi committed:
Delete phase04/bin/Debug/net5.0 directory

rahaahmadi committed:
Delete FileReader.cs

Parsa2820 commented on phase04/Grade.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace ConsoleApp1
```
<p dir="rtl">
واسه 
namespace 
اسم درست بذارید
</p>

Parsa2820 commented on phase04/IO.cs:
```diff
@@ -0,0 +1,32 @@
+            }
+        }
+
+        public string ToString(Student student, double GPA)
```
اسم تابع رو یه چیز دیگه بذارید

Parsa2820 commented on phase04/IO.cs:
```diff
@@ -0,0 +1,32 @@
+        
+        public void PrintOutput(Dictionary<Student, double> studentsList)
+        {
+            var firstThreeStudents = studentsList.Take(3);
```
خوبه که عدد 3 رو به عنوان ورودی بگیرید و هاردکد نباشه

Parsa2820 commented on phase04/JsonConverter.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public List<T> DeserializeObject<T>(string jsonFileContent)
+        {
+            return JsonSerializer.Deserialize<List<T>>(jsonFileContent);
```
<p dir="rtl">
واسه همین یه خط کلاس نوشتید؟ اسم تابع گمراه کنندس. تایپ T ه ولی به 
</p>
List &ltT&gt
<p dir="rtl">
 دیسریالایز میکنه
</p>

Parsa2820 commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            var studentsGrades = GetStudentsGrades();
+            var studentAverageMap = new Dictionary<Student, double>();
+            foreach (var student in _studentsList)
```
میتونید با لینک هم بنویسید
var studentAverageMap = _studentsList.ToDictionary(s => s, 
        studentsGrades
            .Where(grade => grade.StudentNumber == student.StudentNumber)
            .Select(grade => grade.Score)
            .Average());

Parsa2820 commented on phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public Dictionary<Student, double> GetSortedGPAs()
+        {
+            return SetStudentAverage().OrderByDescending(x => x.Value).ToDictionary(x => x.Key, x => x.Value);
```
ToDictionary 
واسه چی زدید؟

rahaahmadi replied:
خروجیش
IOrderedEnumerable 
بود میخواستیم دیکشنری باشه

Parsa2820 reviewed (commented):
خوب بود. خسته نباشید

rahaahmadi committed:
Add files via upload

rahaahmadi committed:
Create .gitignore

rahaahmadi committed:
Delete Phase04/Phase04 directory

rahaahmadi committed:
Add files

rahaahmadi committed:
Delete phase04 directory

hoseinaghaei commented on Phase04/Phase04/IO.cs:
```diff
@@ -0,0 +1,32 @@
+            }
+        }
+
+        private string OutputString(Student student, double GPA)
```
اسم تابع خوب نیست

hoseinaghaei replied:
اینم بگم، حالا اینجا شی
student
کلا ۳ تا فیلد دارع و شما دوتاش نیاز داشتین تو این تابع
اما اگر یک جایی فیلد شی بیشتر باشع و شما هم به یکی دوتا فیلدش نیاز داشته باشین باید فیلدهاش پاس بدین نه خودشو

hoseinaghaei commented on Phase04/Phase04/IO.cs:
```diff
@@ -0,0 +1,32 @@
+            return fileContent;
+        }
+        
+        public void PrintOutput(Dictionary<Student, double> studentsList, int numberOfStudentToPrint)
```
اینکه کل پراسس رو انجام بدیم روی همه داده ها و تهش موقع پرینت تعداد مشخص کنیم کار خوبی نیست
شما میتونی موقع پراسس تعداد رو تاثیر بدی

hoseinaghaei commented on Phase04/Phase04/Program.cs:
```diff
@@ -0,0 +1,16 @@
+            var io = new IO();
+            var jsonConverter = new JsonConverter();
+            var studentsList = jsonConverter.GetDeserializedObjects<Student>(io.GetInput("JsonData/Students.json"));
+            var scoresList = jsonConverter.GetDeserializedObjects<Grade>(io.GetInput("JsonData/Scores.json"));
```
ادرس هارو کانست بالای تابع یا کلاس تعریف کنید

hoseinaghaei commented on Phase04/Phase04/JsonConverter.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public List<T> GetDeserializedObjects<T>(string jsonFileContent)
+        {
+            return JsonSerializer.Deserialize<List<T>>(jsonFileContent);
```
بهتره تو کار با لیست های بزرگ و این حرفا بجای لیست از 
IEnumrable
استفاده کنید
چون شاید اصلا نیاز نشه که روی همش کوئری بزنیم

hoseinaghaei commented on Phase04/Phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,37 @@
+
+        private List<Grade> GetStudentsGrades()
+        {
+            return _studentsList.Join(_scoresList, std => std.StudentNumber, grade => grade.StudentNumber, (std, grade) => grade).ToList();
```
چون یک کار تقریبا پیچیده داره انجام میشه بهتره اول بریزین تو ی متغییر که با اسمش کسی که داره کد رو میخونه بفهمه برای چی زدینش و بعد اونو ریترن کنین

hoseinaghaei commented on Phase04/Phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,37 @@
+            return _studentsList.Join(_scoresList, std => std.StudentNumber, grade => grade.StudentNumber, (std, grade) => grade).ToList();
+        }
+
+        private Dictionary<Student, double> SetStudentAverage()
```
تابع ست خروجی نداره
اسم بهتر برای تابع

hoseinaghaei commented on Phase04/Phase04/ScoresOrganizer.cs:
```diff
@@ -0,0 +1,37 @@
+                new {
+                Student = s,
+                Average = studentsGrades.Where(p => p.StudentNumber == s.StudentNumber).Select(x => x.Score).Average()
+                 }).ToDictionary(x => x.Student, x => x.Average);
```
اینجا اول یک شی
Anonymous 
ساختین بعد تبدیل کردین هردو فیلدش رو به دیکشنری
خب از اول دیکشنری بسازین
یا کلا یک کلاس بسازین که تو خودش 
student & average
داره و اونو ریترن کنین

rahaahmadi committed:
Added constant file path

rahaahmadi committed:
Update

Melikad committed:
applied number of students to print to process part
changed naming

Melikad committed:
changed naming
changed List to IEnumerable
merged getStudentsGrades and getStudentsAverage into one method using GroupJoin and a new class StudentAverage
